### ISSUE TITLE
(apigateway): support deployment salting for imported RestApi

### ISSUE THREAD

**Issue by taoatmars** [aws/aws-cdk#12417]
If the RestApi is define by RestApi.fromRestApiAttributes(), any deployment will not add salt to the deployment resource.

### Reproduction Steps
This is my code.
```
const api = RestApi.fromRestApiAttributes(this, 'ImportedApi', {
      restApiId: XXX,
      rootResourceId: XXX,
    });

    const allResources = api.root.addResource('cdk');

    allResources.addCorsPreflight({ allowOrigins: Cors.ALL_ORIGINS });

    const method = allResources.addMethod('PUT', ifn, {
      methodResponses: [
        {
          statusCode: '200',
          responseParameters: {
            'method.response.header.Access-Control-Allow-Origin': true,
          },
        },
      ],
      authorizationType: AuthorizationType.IAM,
    });

    // Handle deployment
    const deployment = new Deployment(this, '123_deployment', {
      api: RestApi.fromRestApiId(this, 'abc', apiCDKApiApiId),
    });
    new Stage(this, 'CDKApiStageDev', { deployment, stageName: `${ENV}` });
```

"taotest123deploymentC642C343" is the resource name in the Cloudformation template.

### What did you expect to happen?

add salt to the resource name

### What actually happened?

no salt was added to the resource name

### Environment

  - **CDK CLI Version  :*1.83*
  - **Framework Version:**
  - **Node.js Version:** 
  - **OS               :*MAC*
  - **Language (Version):*TypeScript* 
  
### Other

**WORKAROUND**: Just in case anyone ran into the same problem, you can force a deployment by manually adding salt to deployment id. in above example, 123_deployment is my deployment id, you can change the id 1234_deployment to force new deployment to deploy. 

--- 

This is :bug: Bug Report

**Comment by nija-at** [comment]
Unfortunately, this isn't easy to implement since import APIs are intended to create API Gateway resources that are unaware of others. This is how we are able to split off these resources into different stacks and not cause cyclic dependencies.

Manually salting the deployment (as described in your workaround) is the best option available.

Moving this to a feature request.

**Comment by taoatmars** [comment]
@nija-at Thanks for the reply, I know this is difficult to implement, but

**Comment by tmclaugh** [comment]
Per https://github.com/aws/aws-cdk/issues/13383, if _latestDeployment_ were to be set when calling _RestApi.fromRestApiAttributes()_ this issue should be resolved without even needing to create a Deployment resource manually within the stack.

**Comment by nija-at** [comment]
Hi @tmclaugh - I'm keen to explore your statement here.

The [`addToLogicalId()` method](https://github.com/aws/aws-cdk/blob/9d228e4ba10ebf0d5a5bb05d3fc172be19bb3fd5/packages/%40aws-cdk/aws-apigateway/lib/deployment.ts#L100) is the key to determine when a `Deployment`'s logical id should be modified. You can see [all the uses of this method](https://github.com/aws/aws-cdk/search?q=addToLogicalId) that tells the Deployment when it should be updated.

Let's take the simple case where a `RestApi` is defined in one stack along with a number of Resources and Methods, and later imported into another stack using the `RestApi.fromRestApiAttributes()` where further Methods and Resources are defined. 
Unless `deploy` is unset, the first stack would create a `Deployment` construct set to `latestDeployment`. However, it would not be possible for this `Deployment` construct to know of the Methods and Resources in the second stack and therefore cannot correctly determine when its logical id should be mutated.

How do you propose your approach tackle this case?

**Comment by montoyan877** [comment]
Any news on this topic?

**Comment by Durgaprasad-Budhwani** [comment]
Hey Team,

I am using the below code which is working for me.

```js

    const deployment = new Deployment(this, `deployment-${new Date().toISOString()}`, { api: apiGateway });

    // @ts-ignore
    deployment.resource.stageName = ENV!;

    const logicalId = Date.now().toString();// salted id
    const cfnDeployment = deployment.node.defaultChild as CfnDeployment;
    cfnDeployment.overrideLogicalId(logicalId);
    
```

I am doing deployment two times to make it work.
In the first deployment, it s is resources and in second deployment, it is updating stage.

This will be like 

```bash

          npx aws-cdk synth
          npx aws-cdk deploy --require-approval never
          npx aws-cdk deploy --require-approval never       
```

**Comment by nicks6853** [comment]
+1 to this. It's the only reason that is preventing us from using AWS CDK for projects at work. Creating a new deployment every time by changing the deployment id seems overkill.

**Comment by otaviomacedo** [comment]
I don't see how we can solve this without doing hacks. The core problem, as @nija-at explained, is that, by importing a Rest API, you don't have access to its Deployment.

I'm closing this issue for now, but if someone has an idea for a solution, please share it with us, either by creating a new ticket or submitting a PR, and we may come back to this.

**Comment by robert-pitt-foodhub** [comment]
I overcome this by overwriting the _latestDeployment variable on the RestApi, this needs to be done prior to adding any methods.

```
  //...
  /**
   * Initialise the rest api
   */
  public readonly restApi = new RestApi(this, 'api', {
    restApiName: `falcon-${this.props.envName}-poshub`,
    deploy: false,
    defaultCorsPreflightOptions,
  })

  /**
   * Documentation Version
   */
  public readonly version = new CfnDocumentationVersion(this, 'version', {
    documentationVersion: '1.1.44',
    restApiId: this.restApi.restApiId,
    description: 'this is a test of documentation',
  })

  /**
   * Rest API Deployment
   */
  public readonly deployment = new Deployment(this, 'deployment', {
    api: this.restApi,
    description: `Deployment: ${this.props.envName}/${this.version.documentationVersion}`,
  })

  // This patch needs to be applied prior to mounting the ApiMethods construct
  // This being present on the restApi allows for all methods to be bound to the deployment
  protected forceLatestDeploymentOverride = (this.restApi['_latestDeployment'] =
    this.deployment)
  //...
    ```

This may work for imported reset api's as well but I haven't tested, this resolves two issues with my setup:

1. Ensuring that the all the methods created in my API effect the deployments logical id to ensure redeployment when methods change
2. Ensuring that all methods are bound to the deployment as a dependancy, this will ensure that my post deployment constructs to extract documetation run against the very latest deployment.

Hope this helps
